### PR TITLE
Support CLCACHE_BASEDIR in nodirect mode

### PR DESCRIPTION
**Pull request by oktal3700**
Do a simple case-insensitive find-and-replace to transform absolute
paths into relative paths within the preprocessor output that is used
to compute the hash. This makes CLCACHE_NODIRECT mode usable in the
presence of the `__FILE__` macro.

Fixes a bug in _normalizedCommandLine that caused the source filename
to be included in the hash computation.

### CODE REVIEW

**Comment by oktal3700** [comment]
Not sure what caused the build to fail.

File: clcache/__main__.py [hunk 1]
```diff
@@ -118,6 +118,19 @@ def normalizeBaseDir(baseDir):
...
+    output = b''
+    index = 0
+    while True:
+        nextIndex = dataLower.find(oldLower, index)
+        if nextIndex < 0:
+            return output + data[index:]
+        output += data[index:nextIndex] + new
+        index = nextIndex + len(old)
+
+
```
Comment (frerich) [inline clcache/__main__.py]: I'm not sure you need a dedicated function for this -- maybe it would be viable to just use `re.sub` here?  I think `caseInsensitiveReplace` is basically

```python
re.sub(re.escape(old), lambda _: new, data, flags=re.IGNORECASE)
```

File: clcache/__main__.py [hunk 2]
```diff
@@ -469,6 +482,11 @@ def computeKeyNodirect(compilerBinary, commandLine, environment):
         compilerHash = getCompilerHash(compilerBinary)
         normalizedCmdLine = CompilerArtifactsRepository._normalizedCommandLine(commandLine)
 
+        if "CLCACHE_BASEDIR" in os.environ:
+            baseDir = normalizeBaseDir(os.environ["CLCACHE_BASEDIR"]).replace("\\", "\\\\").encode("UTF-8")
```
Comment (frerich) [inline clcache/__main__.py:473]: Hmm, why are backslashes escaped here?

File: clcache/__main__.py [hunk 3]
```diff
@@ -495,7 +513,7 @@ def _normalizedCommandLine(cmdline):
         argsToStrip += ("MP",)
 
         return [arg for arg in cmdline
-                if not (arg[0] in "/-" and arg[1:].startswith(argsToStrip))]
+                if arg[0] in "/-" and not arg[1:].startswith(argsToStrip)]
```
Comment (frerich) [inline clcache/__main__.py:503]: Hmm, did you mean to change the behaviour here? I suspect the `and` should become `or` here. [De Morgan](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) found that `not (a and b)` is equivalent to `not a or not b`.

That being said, I'm not sure the new version is any nicer than the old one, so maybe it's not worth touching this at all.

File: tests/test_unit.py [hunk 4]
```diff
@@ -109,6 +109,13 @@ def testFilesBeneathRecursive(self):
             self.assertIn(r".\d\4.txt", files)
             self.assertIn(r".\d\e\5.txt", files)
 
+    def testCaseInsensitiveReplace(self):
+        replace = lambda s: clcache.caseInsensitiveReplace(s, b"PaTtErN", b"replacement")
+        self.assertEqual(replace(b"pattern may appear"), b"replacement may appear")
+        self.assertEqual(replace(b"infix paTTErn embed"), b"infix replacement embed")
+        self.assertEqual(replace(b"Pattern and PATTERN and pattern"), b"replacement and replacement and replacement")
+        self.assertEqual(replace(b"and finally PATTERN"), b"and finally replacement")
+
```
Comment (frerich) [inline tests/test_unit.py]: I suppose that in case you decide to drop the custom `caseInsensitiveReplace` function in favor of `re.sub`, you could also drop this test.

**Review by frerich — COMMENTED** [review]
I think this looks like a plausible improvement, thanks a lot! However, it also invalidates all existing cache entries for people who do _not_ use the direct mode, i.e. I guess this shouldn't be part of a bugfix release but rather of a feature release.

File: clcache/__main__.py (hunk 3 above)
Comment (frerich) [reply clcache/__main__.py:503]: Oh, I only now noticed that you _meant_ to fix a bug here. In that case, can you please add a test case which demonstrates the bug?

File: clcache/__main__.py (hunk 1 above)
Comment (oktal3700) [reply clcache/__main__.py]: Rightly or wrongly, for simple things I personally like to avoid regex for performance reasons. But I'm a beginner in Python so if you prefer regex here I'm happy to oblige. And `re.escape` should protect it from any shenanigans.

File: clcache/__main__.py (hunk 2 above)
Comment (oktal3700) [reply clcache/__main__.py:473]: It's to match the backslashes that are escaped in the preprocessor output because `__FILE__` expands to a C string literal.

File: clcache/__main__.py (hunk 3 above)
Comment (oktal3700) [reply clcache/__main__.py:503]: `testBasedirNoDirect` would fail without this bugfix, but I can add a test that exercises _only_ this bugfix and none of the other changes, if you like.

**Comment by oktal3700** [comment]
> it also invalidates all existing cache entries for people who do _not_ use the direct mode

Are you sure? Only `computeKeyNodirect` is affected. The bugfix part will invalidate cache entries for people who do not use `BASEDIR` though, is that what you meant?

**Comment by oktal3700** [comment]
Applied suggested changes.
